rfc/rfc-64/rfc-64.md

```diff
@@ -0,0 +1,509 @@
+    // Ordered list of the delta-log files 
+    TreeSet<HoodieLogFile> logFilesPaths;
+    // ...
+}
+```
+
+NOTE: In case of COW set of records within the File Group wouldn't change.
+
+```java
+// NOTE: This model is the same as an existing one (ref'd here for notational purposes)
+class HoodieFileGroup {
+    // Id of the file-group
+    HoodieFileGroupId id;
+    // Projection of the timeline of committed actions affecting file-slices in this group
+    HoodieTimeline timeline;
+    // File-slices ordered by corresponding instant they've been committed at
+    TreeMap<HoodieInstant, FileSlice> fileSlices;
+    // ...
+}
+```
+
+**PartitionDescriptor** describes particular partition within the dataset.
+
+```java
+class PartitionDescriptor {
+    // Relative partition path within the dataset
+    String partitionPath;
+    // List of values for corresponding partition columns (partition values)
+    List<Object> partitionColumnValues;
+}
+```
+
+**PartitionSnapshot** represents a state of a single table's partition at any
+particular instant T, comprised of the file-slices that are the _latest_ (
+current) within their corresponding file-groups at the instant T.
+
+NOTE: Non-partitioned tables assumed to have 1 partition enclosing the whole
+table.
+
+```java
+class PartitionSnapshot {
+    PartitionDescriptor descriptor;
+    // (Latest) Instant T as of which partition's state is represented
+    HoodieInstant latestInstant;
+    // File-slices current at the instant T
+    List<HoodieFileSlice> fileSlices;
+}
+```
+
+**PartitionIncrementalSnapshot** represents a state of a single table's
+partition at a particular instant T2, while only considering commits that
+occurred in the table after T1 (ie within the timeline T1 -> T2)
+comprised of the file-slices that are the _latest_ (current) within their
+corresponding file-groups at instant T2.
+
+NOTE: `PartitionSnapshot` is a special case of `PartitionIncrementalSnapshot` where
+T1 = 0.
+
+```java
+class PartitionIncrementalSnapshot {
+    PartitionDescriptor descriptor;
+    // Timeline starting at T1 and ending at T2, completed actions thereof 
+    // represent state of the partition
+    HoodieTimeline timeline;
+    // File-slices current at the instant T_y
+    List<HoodieFileSlice> fileSlices;
+}
+```
+
+**ReadingMode** represents the query type.
+
+```java
+enum ReadingMode {
+    // Representing state of the table at a particular instant T
+    SNAPSHOT,
+    // (MOR only) Representing state of the table at a particular instant T,
+    // reading exclusively base-files (ie, ignoring any delta-logs if present)
+    SNAPSHOT_READ_OPTIMIZED,
+    // Representing state of the table at a particular instant T_y assuming,
+    // as an initial instant Tx (snapshot is a specification of this mode, where T_x = 0)
+    INCREMENTAL,
+    // Representing state of the table as a stream of CRUD operations on
+    // individual records
+    CDC
+}
+```
+
+**Expression** (**Predicate**)
+
```

Comment:
only use for dataSkipping/partition prune ？
 
we donnot think we should use those predicate for engine scan action。

Reply:
Correct this will only be used for data filtering (skipping, partition-pruning, filters push-down)